src/main/java/hudson/plugins/ec2/EC2AbstractSlave.java

```diff
@@ -508,14 +508,14 @@ void stop() {
 
     }
 
-    protected void cleanupSshKeyPairIfNeeded() {
-        if (getInstanceSshKeyPairName() != null) {
+    protected void cleanupSshKeyPairIfNeeded() throws java.io.IOException {
+        if (getCloud().resolveKeyPair() == null) {
```

Comment:
Now we are always stashing a reference to the keypair regardless of using dynamic or static keys, so you can't rely on the prescense of this field to tell you how the controller is configured.

Reply:
Seems fragile. I would expect the serialized state of the agent to indicate whether it has an ephemeral keypair, and if so, its name. Otherwise you are back to the situation where switching from one mode to the other could result in generated keypairs not being deleted automatically. I thought this had been fixed at some point earlier in this PR’s history but it is now regressed?

Reply:
i think its the result of the changes i did to always store the key  (regardless of using dynamic keys) - i should be able to unwind that and resolve this comment